### PR TITLE
Support adding an environment to a workspace.

### LINKED ISSUE #4852: Open Folder: Add environment dialog should use and update folder settings

### PR DESCRIPTION
**Pull request by huguesv**
Add environment dialog is now aware of workspaces (defaults to workspace folder, reads and updates the pythonsettings.json)

Fix #4852

### CODE REVIEW

File: Common/Tests/Utilities.UI/UI/SolutionExplorerTree.cs [hunk 1]
```diff
@@ -125,6 +125,40 @@ public SolutionExplorerTree(AutomationElement element)
...
+
+        public TreeNode WaitForChildOfWorkspace(TimeSpan timeout, params string[] path) {
+            var item = WaitForItemHelper(p => FindChildOfWorkspaceHelper(p, false), path, timeout);
+            // Check one more time, but now let the assertions be raised.
+            return new TreeNode(FindChildOfWorkspaceHelper(path, true));
+        }
+
+        private AutomationElement FindChildOfWorkspaceHelper(string[] path, bool assertOnFailure) {
+            var projElement = Nodes.First()?.Element;
+            if (assertOnFailure) {
```
Comment (MikhailArkhipov) [inline Common/Tests/Utilities.UI/UI/SolutionExplorerTree.cs]: FirstOrDefault or nodes are always non empty?

File: Python/Product/PythonTools/PythonTools/Environments/AddEnvironmentDialog.xaml.cs [hunk 2]
```diff
@@ -143,19 +154,27 @@ enum PageKind {
...
+            if (workspace != null) {
+                var registryService = site.GetComponentModel().GetService<IInterpreterRegistryService>();
+                var optionsService = site.GetComponentModel().GetService<IInterpreterOptionsService>();
+                var factory = workspace.GetInterpreterFactory(registryService, optionsService);
+                selectedProjectView = new ProjectView(workspace, factory);
+                projectViews = new ProjectView[] { selectedProjectView };
+            } else {
+                try {
+                    var sln = (IVsSolution)site.GetService(typeof(SVsSolution));
+                    var projects = sln?.EnumerateLoadedPythonProjects().ToArray() ?? new PythonProjectNode[0];
```
Comment (MikhailArkhipov) [inline Python/Product/PythonTools/PythonTools/Environments/AddEnvironmentDialog.xaml.cs]: Array.Empty<>

Comment (MikhailArkhipov) [reply Python/Product/PythonTools/PythonTools/Environments/AddEnvironmentDialog.xaml.cs]: It sure what happens when null.ToArray

File: Python/Product/PythonTools/PythonTools/Environments/VirtualEnv.cs [hunk 3]
```diff
@@ -140,34 +159,37 @@ static class VirtualEnv {
...
-                        output.ShowAndActivate();
-                    } else {
-                        output.Show();
-                    }
-                }
 
-                if (exitCode != 0 || !Directory.Exists(path)) {
-                    throw new InvalidOperationException(Strings.VirtualEnvCreationFailed.FormatUI(path));
+                    if (exitCode != 0 || !Directory.Exists(path)) {
+                        throw new InvalidOperationException(Strings.VirtualEnvCreationFailed.FormatUI(path));
```
Comment (MikhailArkhipov) [inline Python/Product/PythonTools/PythonTools/Environments/VirtualEnv.cs]: Does it show something useful to the user, such as message box?

File: Python/Product/PythonTools/PythonTools/Environments/AddExistingEnvironmentView.cs [hunk 4]
```diff
@@ -508,24 +508,32 @@ ProjectView selectedProject
...
-                        RegistryService,
-                        PrefixPath,
-                        InterpreterPath,
-                        WindowsInterpreterPath,
-                        PathEnvironmentVariable,
-                        version,
-                        InterpreterArchitecture.TryParse(ArchitectureName ?? ""),
-                        Description
-                    );
+                    if (SelectedProject.Node != null) {
```
Comment (MikhailArkhipov) [inline Python/Product/PythonTools/PythonTools/Environments/AddExistingEnvironmentView.cs]: Some other places also check SelectedProject for null.

File: Python/Product/PythonTools/PythonTools/Environments/AddEnvironmentDialog.xaml.cs (hunk 2 above)
Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/Environments/AddEnvironmentDialog.xaml.cs]: if `snl?` is `null`, it should stop evaluating that expression and go to the `??` part

File: Python/Product/PythonTools/PythonTools/Environments/VirtualEnv.cs (hunk 3 above)
Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/Environments/VirtualEnv.cs]: Yes, in output window.

File: Python/Product/PythonTools/PythonTools/Environments/AddExistingEnvironmentView.cs (hunk 4 above)
Comment (huguesv) [reply Python/Product/PythonTools/PythonTools/Environments/AddExistingEnvironmentView.cs]: We're inside a `if (SelectedProject != null)` here.

**Pull request merged**